### PR TITLE
[IMP] sale: performance

### PR DESCRIPTION
**Pull request by Feyensv**
For big Sales Order, onchanges request may take up to multiple seconds.

1) Enable prefetch in batch for NewIds: https://github.com/odoo/odoo/pull/45959
2) Fix the computes and/or onchanges taking way too much time because of bad design/...

> `invoice_status`: do not compute the invoice_status for unconfirmed SO, it is `no` anyway, no need to iterate on all lines
>  `virtual_available_at_dates, ...`: those sale_stock fields are computed information fields for the stock widget, but an opportunity to batch them was missed.
> `qty_invoiced`: For invoiced Sales Orders, the qty_invoiced computation was taking a lot of time, considering all invoice lines information.  This should be done directly in SQL (0.5s gain for 70+ SO lines).
> `amount_delivery`: this field is computed, even if not present in the view, during the final flush of the request. No big improvement can be done in stable, but the compute method is a bit cleaned up.

[Task Id-2198601](https://www.odoo.com/web?#id=2198601&action=333&active_id=2818&model=project.task&view_type=form&cids=1&menu_id=4720)
opw-2186113

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_stock/models/sale_order.py [hunk 1]
```diff
@@ -235,15 +235,15 @@ def _compute_qty_at_date(self):
...
         # date in order to batch the read of the quantities computed field.
+        now = fields.Datetime.now()
         for line in self:
             if not line.display_qty_widget:
                 continue
             line.warehouse_id = line.order_id.warehouse_id
             if line.order_id.commitment_date:
                 date = line.order_id.commitment_date
             else:
-                confirm_date = line.order_id.date_order if line.order_id.state in ['sale', 'done'] else datetime.now()
```
Comment (Feyensv) [inline addons/sale_stock/models/sale_order.py]: After consideration, still useful for new lines in confirmed SO's, to verify and bring back

File: addons/stock/models/product.py [hunk 2]
```diff
@@ -94,8 +94,8 @@ class Product(models.Model):
     @api.depends('stock_move_ids.product_qty', 'stock_move_ids.state')
     @api.depends_context(
         'lot_id', 'owner_id', 'package_id', 'from_date', 'to_date',
-        'company_owned', 'force_company',
-    )
+        'company_owned', 'force_company', 'warehouse',
+    ) # warehouse context key is needed for sale_stock widget, reading available quantities in batch of warehouse,to_date contexts
```
Comment (Feyensv) [inline addons/stock/models/product.py]: re drop

File: addons/delivery/models/sale_order.py [hunk 3]
```diff
@@ -27,11 +27,8 @@ def _compute_amount_total_without_delivery(self):
 
     @api.depends('order_line')
     def _compute_delivery_state(self):
-        delivery_line = self.order_line.filtered('is_delivery')
-        if delivery_line:
-            self.delivery_set = True
-        else:
-            self.delivery_set = False
+        for order in self:
+            order.delivery_set = bool(order.order_line.filtered('is_delivery'))
```
Comment (Feyensv) [inline addons/delivery/models/sale_order.py]: if any

File: addons/delivery/models/sale_order.py [hunk 4]
```diff
@@ -126,7 +123,7 @@ def _format_currency_amount(self, amount):
                  'order_line.product_id.invoice_policy')
     def _get_invoice_status(self):
         super()._get_invoice_status()
-        for order in self:
+        for order in self.filtered(lambda order: order.invoice_status not in ['no', 'invoiced']):
```
Comment (Feyensv) [inline addons/delivery/models/sale_order.py]: if ...:
continue

File: addons/sale/models/sale.py [hunk 5]
```diff
@@ -73,21 +73,31 @@ def _get_invoice_status(self):
...
-            elif line_invoice_status and all(invoice_status == 'invoiced' for invoice_status in line_invoice_status):
-                order.invoice_status = 'invoiced'
-            elif line_invoice_status and all(invoice_status in ('invoiced', 'upselling') for invoice_status in line_invoice_status):
-                order.invoice_status = 'upselling'
-            else:
-                order.invoice_status = 'no'
+        unconfirmed_orders = self.filtered(lambda so: so.state not in ['sale', 'done'])
+        unconfirmed_orders.invoice_status = 'no'
+        confirmed_orders = self - unconfirmed_orders
+        if confirmed_orders:
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: if not confirmed_orders:
return

for less diff

File: addons/sale/models/sale.py [hunk 6]
```diff
@@ -1025,23 +1035,37 @@ def _get_to_invoice_qty(self):
...
+                    SUM(CASE WHEN am.type = 'out_refund' THEN aml.quantity / aml_uom.factor * sol_uom.factor ELSE 0 END) as out_refund
+                FROM sale_order_line_invoice_rel rel
+                JOIN sale_order_line sol on sol.id = rel.order_line_id
+                JOIN account_move_line aml on aml.id = rel.invoice_line_id
+                JOIN account_move am on am.id = aml.move_id
+                JOIN uom_uom sol_uom on (sol_uom.id=sol.product_uom)
+                JOIN uom_uom aml_uom on (aml_uom.id=aml.product_uom_id)
+                WHERE rel.order_line_id IN %s
+                GROUP BY rel.order_line_id
+            ''', [tuple(self.ids)])
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: existing_lines.ids

File: addons/sale/models/sale.py [hunk 7]
```diff
@@ -1025,23 +1035,37 @@ def _get_to_invoice_qty(self):
...
-    @api.depends('invoice_lines.move_id.state', 'invoice_lines.quantity')
+    @api.depends('product_uom', 'invoice_lines.move_id.state', 'invoice_lines.quantity', 'invoice_lines.product_uom_id')
     def _get_invoice_qty(self):
         """
         Compute the quantity invoiced. If case of a refund, the quantity invoiced is decreased. Note
         that this is the case only if the refund is generated from the SO and that is intentional: if
         a refund made would automatically decrease the invoiced quantity, then there is a risk of reinvoicing
         it automatically, which may not be wanted at all. That's why the refund has to be created from the SO
         """
+        existing_lines = self.filtered('id')
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: `origin_lines = self._origin`
if origin_lines:...

origin_lines.ids

File: addons/sale_stock/tests/test_sale_stock.py [hunk 8]
```diff
@@ -645,6 +645,9 @@ def test_09_qty_available(self):
         self.assertEqual(line.warehouse_id, warehouse1)
         self.assertEqual(line.qty_to_deliver, 1)
         so.warehouse_id = warehouse2
+        # invalidate product cache to ensure qty_available is recomputed
+        # bc warehouse isn't in the depends_context of qty_available
+        line.product_id.invalidate_cache()
```
Comment (Feyensv) [inline addons/sale_stock/tests/test_sale_stock.py]: add flush before
and invalidate only targeted fnames

File: addons/sale/models/sale.py [hunk 9]
```diff
@@ -1025,23 +1036,37 @@ def _get_to_invoice_qty(self):
...
+            ''', [tuple(origin_lines.ids)])
+
+            query_res = self._cr.fetchall()
+        else:
+            query_res = []
+        results = {res[0]: res[1]-res[2] for res in query_res}
         for line in self:
-            qty_invoiced = 0.0
-            for invoice_line in line.invoice_lines:
-                if invoice_line.move_id.state != 'cancel':
```
Comment (dbo-odoo) [inline addons/sale/models/sale.py]: you've lost this condition in your query AFAICT

File: addons/sale/models/sale.py [hunk 10]
```diff
@@ -1025,23 +1036,37 @@ def _get_to_invoice_qty(self):
...
         a refund made would automatically decrease the invoiced quantity, then there is a risk of reinvoicing
         it automatically, which may not be wanted at all. That's why the refund has to be created from the SO
         """
+        origin_lines = self._origin
+        if origin_lines:
+            self.env.cr.execute('''
+                SELECT
+                    rel.order_line_id as id,
+                    SUM(CASE WHEN am.type = 'out_invoice' THEN aml.quantity / aml_uom.factor * sol_uom.factor ELSE 0 END) as out_invoice,
+                    SUM(CASE WHEN am.type = 'out_refund' THEN aml.quantity / aml_uom.factor * sol_uom.factor ELSE 0 END) as out_refund
```
Comment (dbo-odoo) [inline addons/sale/models/sale.py]: missing a rounding when comparing to the `_compute_quantity` method of `uom.uom`

File: addons/website_sale_delivery/models/sale_order.py [hunk 11]
```diff
@@ -14,19 +14,22 @@ class SaleOrder(models.Model):
...
+    # VFE TODO this field help should be updated, bc the amount can be with tax depending
+    # on the current user groups.
 
     def _compute_website_order_line(self):
         super(SaleOrder, self)._compute_website_order_line()
         for order in self:
             order.website_order_line = order.website_order_line.filtered(lambda l: not l.is_delivery)
 
-    @api.depends('order_line.price_unit', 'order_line.tax_id', 'order_line.discount', 'order_line.product_uom_qty')
+    @api.depends('order_line.price_subtotal')
```
Comment (dbo-odoo) [inline addons/website_sale_delivery/models/sale_order.py]: I know it's the same compute method, so if one changes so will the other; but i'd prefer to explicitely have `price_total` in the deps as well

File: addons/website_sale_delivery/models/sale_order.py [hunk 12]
```diff
@@ -14,19 +14,22 @@ class SaleOrder(models.Model):
         compute='_compute_amount_delivery',
         string='Delivery Amount',
         help="The amount without tax.", store=True, tracking=True)
+    # VFE TODO this field help should be updated, bc the amount can be with tax depending
```
Comment (dbo-odoo) [inline addons/website_sale_delivery/models/sale_order.py]: no_todo

Comment (Feyensv) [reply addons/website_sale_delivery/models/sale_order.py]: TODO for master (and with PO's advise), I do not want to update this in 13.0 :P.
The field is only shown in mail templates and the e-commerce so unless studio is used, it isn't shown as a field.

Comment (dbo-odoo) [reply addons/website_sale_delivery/models/sale_order.py]: create a task or a PR for master then :p

**Comment by Feyensv** [comment]
@robodoo r+ rebase-ff

File: addons/sale/models/sale.py [hunk 13]
```diff
@@ -1025,23 +1034,48 @@ def _get_to_invoice_qty(self):
...
         that this is the case only if the refund is generated from the SO and that is intentional: if
         a refund made would automatically decrease the invoiced quantity, then there is a risk of reinvoicing
         it automatically, which may not be wanted at all. That's why the refund has to be created from the SO
         """
+        origin_lines = self._origin
+        if origin_lines:
+            # PSQL ROUND does round sometimes down, which isn't the rounding_method we want.
+            # We therefore use CEIL, but as CEIL doesn't support decimal rounding,
+            # we divide by the uom rounding, before ceiling and multiply after to preserve the decimal places.
+            # (and the default rounding method of UOM _compute_quantity)
```
Comment (antonylesuisse) [inline addons/sale/models/sale.py]: Test ?

**Comment by tbe-odoo** [comment]
Hello guys, this breaks the rounding for the quantities.
Here we cannot easily implement the rounding made by `float_round` on decimals and without taking in account the epsilon as done in the python code.

Here is an example, using `CEIL` completely breaks the value we wait for

```sql
SELECT
    rel.order_line_id as id,
    aml.quantity / aml_uom.factor * sol_uom.factor / sol_uom.rounding,
    SUM(CASE WHEN am.type = 'out_invoice' THEN
        CEIL(aml.quantity / aml_uom.factor * sol_uom.factor / sol_uom.rounding) * sol_uom.rounding
        ELSE 0 END) as out_invoice,
    SUM(CASE WHEN am.type = 'out_refund' THEN
        CEIL(aml.quantity / aml_uom.factor * sol_uom.factor / sol_uom.rounding) * sol_uom.rounding
        ELSE 0 END) as out_refund
FROM sale_order_line_invoice_rel rel
JOIN sale_order_line sol on sol.id = rel.order_line_id
JOIN account_move_line aml on aml.id = rel.invoice_line_id
JOIN account_move am on am.id = aml.move_id
JOIN uom_uom sol_uom on (sol_uom.id=sol.product_uom)
JOIN uom_uom aml_uom on (aml_uom.id=aml.product_uom_id)
WHERE rel.order_line_id IN (2765394) AND am.state != 'cancel'
GROUP BY rel.order_line_id, 2;
```

```
   id    |               ?column?                | out_invoice | out_refund 
---------+---------------------------------------+-------------+------------
 2765394 | 13.0000000000000000003749999999999867 |        14.0 |          0
```

I have to revert this commit until we fix this issue.

**Pull request closed without merging**